### PR TITLE
fix: Considers strokeAlign to compute border dimensions

### LINKED ISSUE #155491: Container border have wrong dimensions when BorderSide width are same but different strokeAlign
### Steps to reproduce

1. Run the provided code sample.
2. Click the container to trigger the animation and see the text jump.

### Expected results

Child widget should be centered like this :

<details open>
<summary>Screenshot - Expected result</summary>

</details>

### Actual results

Child widget a above the center when all `BorderSide` have the same width. With the video we can see the "jump" as soon as widths become the same.

<details open>
<summary>Screenshot - Actual result</summary>

</details>

Note this behavior occurs only with `Container` but not with `DecoratedBox`.

Actually the child moves because of `Container.padding` which depends on its decoration padding :

https://github.com/flutter/flutter/blob/2f149da67614349cb60bc77b465ce6a533aa951e/packages/flutter/lib/src/widgets/container.dart#L371-L377

`BoxDecoration.padding` depends on the `Border.dimensions` :

https://github.com/flutter/flutter/blob/2f149da67614349cb60bc77b465ce6a533aa951e/packages/flutter/lib/src/painting/box_decoration.dart#L215-L216

Defined in `BoxBorder.dart` like :

https://github.com/flutter/flutter/blob/2f149da67614349cb60bc77b465ce6a533aa951e/packages/flutter/lib/src/painting/box_border.dart#L464-L470

From my understanding, at the end of animation when all witdh become the same, `_widthIsUniform = true` and we return `top.strokeInset` for all dimension sides, ignoring `strokeAlign`.

### Ideas to solve

- As `strokeInset` is a product of `width` and also `strokeAlign`, it might be better to also check if strokeAlign is uniform like :

```dart
if (_widthIsUniform && _strokeAlignIsUniform) {
```

Tested locally, it solves this issue but I don't know if it breaks some tests or creates unexpected behavior (maybe ping @\gspencergoog @\bernaferrari as they work a lot on this topic with #102112 )

I would be happy to make my first PR if this solution makes sens 😁 

- A dirty workaround is to set a very slightly different width for the side with a different `strokeAlign`, like `width: 20.00001`

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.dark(),
      home: const Scaffold(
        body: Center(
          child: ContainerBorderBug(),
        ),
      ),
    );
  }
}

class ContainerBorderBug extends StatefulWidget {
  const ContainerBorderBug({super.key});

  @override
  State<ContainerBorderBug> createState() => _ContainerBorderBugState();
}

class _ContainerBorderBugState extends State<ContainerBorderBug> {
  bool _isExpanded = false;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        setState(() => _isExpanded = !_isExpanded);
      },
      child: AnimatedContainer(
        duration: const Duration(milliseconds: 500),
        curve: Curves.easeInOut,
        width: 200,
        height: 100,
        decoration: BoxDecoration(
          color: Colors.blue,
          borderRadius: BorderRadius.circular(12),
          border: Border(
            top: BorderSide(color: Colors.red.withOpacity(0.5), width: 20),
            left: BorderSide(color: Colors.red.withOpacity(0.5), width: 20),
            right: BorderSide(color: Colors.red.withOpacity(0.5), width: 20),
            bottom: BorderSide(
              color: Colors.red.withOpacity(0.5),
              width: _isExpanded ? 40 : 20,
              strokeAlign: BorderSide.strokeAlignOutside,
            ),
          ),
        ),
        child: const Center(child: Text("My Text")),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Video demonstration</summary>

https://github.com/user-attachments/assets/a9a67bf1-667c-4ae1-8a69-4cbe7f11bff5

</details>

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.2, on macOS 15.0 24A335 darwin-arm64, locale fr-FR)
    • Flutter version 3.24.2 on channel stable at
      /Users/romanojw10/Documents/developpement/flutter_arm
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4cf269e36d (3 weeks ago), 2024-09-03 14:30:00 -0700
    • Engine revision a6bd3f1de1
    • Dart version 3.5.2
    • DevTools version 2.37.2

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/romanojw10/Library/Android/sdk
    • Platform android-35, build-tools 34.0.0
    • ANDROID_HOME = /Users/romanojw10/Library/Android/sdk
    • Java binary at: /Users/romanojw10/Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Chromium.app/Contents/MacOS/Chromium

[✓] Android Studio (version 2023.3)
    • Android Studio at /Users/romanojw10/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] IntelliJ IDEA Ultimate Edition (version 2024.1.1)
    • IntelliJ at /Users/romanojw10/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.96.0

[✓] Connected device (4 available)
    • iPhone 16 Pro (mobile)          • 158FE3E3-8761-4E2E-9F46-3567E45228CF • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-18-0 (simulator)
    • macOS (desktop)                 • macos                                • darwin-arm64
      • macOS 15.0 24A335 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad                • darwin
      • macOS 15.0 24A335 darwin-arm64
    • Chrome (web)                    • chrome                               •
      web-javascript • Chromium 112.0.5569.0

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by romaingyh**
Fixes #155491 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

There are 3 golden tests that don't pass on my computer, but they were already not passing before my changes.

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by bernaferrari** [comment]
Good catch. Never thought about this scenario.

**Review by bernaferrari — APPROVED** [review]

File: packages/flutter/lib/src/painting/box_border.dart [hunk 1]
```diff
@@ -463,7 +463,7 @@ class Border extends BoxBorder {
 
   @override
   EdgeInsetsGeometry get dimensions {
-    if (_widthIsUniform) {
+    if (_widthIsUniform && _strokeAlignIsUniform) {
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/painting/box_border.dart:466]: Why do we need to distinguish between the uniform and non-uniform cases? Can we just all use the non-uniform constructors?

Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart:466]: Theoretically it is an optimization so you save 3 getters, each has a 4 step math operation, so you are saving 12 calculations per render. Probably affect nothing, but this is run every time everywhere, so I tried to optimize.

If the code is ugly, we could un-optimize, but I don't know, feels fine either way..

Comment (dkwingsmt) [reply packages/flutter/lib/src/painting/box_border.dart:466]: Rather than whether the code is pretty, I think what's more important is the maintainability. You see, why this bug would happen at all is because we fail to count in all factors that make `left.strokeInset` equal to `top.strokeInset`, which allows us to use `top.strokeInset` for all four. But what if more factors are added to `strokeInset` in the future? We'll need to remember to update this as well. In general, it feels like this "shortcut" relies on implementation details.

In terms of efficiency, I don't think you're saving much either. You'll have to read all factors to fully determine the equality anyway, therefore you can't save reading any getters.

In conclusion, I don't think the performance saved here is worth the risk of maintainability.

Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart:466]: Ok, feel free to do this change @romaingyh

Comment (romaingyh) [reply packages/flutter/lib/src/painting/box_border.dart]: @bernaferrari done :)

**Review by dkwingsmt — APPROVED** [review]
LGTM, thanks for the contribution!

**Comment by bernaferrari** [comment]
Google Testing failing 👀

**Comment by bernaferrari** [comment]
Still failing 👀

**Comment by dkwingsmt** [comment]
Thanks for checking. The first time there was some actual errors but I failed to see the relationship with this PR. The second time there wasn't even a "test result" page. Let me try again.
